app/src/main/java/org/simple/clinic/appupdate/CheckAppUpdateAvailability.kt

```diff
@@ -11,35 +12,64 @@ import io.reactivex.Observable
 import org.simple.clinic.appupdate.AppUpdateState.AppUpdateStateError
 import org.simple.clinic.appupdate.AppUpdateState.DontShowAppUpdate
 import org.simple.clinic.appupdate.AppUpdateState.ShowAppUpdate
+import org.simple.clinic.appupdate.CheckAppUpdateAvailability.AppUpdateInfoTaskResult.Failure
+import org.simple.clinic.appupdate.CheckAppUpdateAvailability.AppUpdateInfoTaskResult.Success
 import javax.inject.Inject
 
 class CheckAppUpdateAvailability @Inject constructor(
     private val appContext: Application,
-    private val config: AppUpdateConfig,
+    private val config: Observable<AppUpdateConfig>,
     private val versionUpdateCheck: (Int, Application, AppUpdateConfig) -> Boolean = isVersionApplicableForUpdate
 ) {
 
   fun listen(): Observable<AppUpdateState> {
```

Comment:
Since `appUpdateCallback()` is a private API, I feel wrapping the result in a sealed class to handle only an exception is  overkill. We can change `appUpdateCallback` to return an `Observable<AppUpdateInfo>` and then add an `onErrorReturn` to the `listen()` chain to handle this without changing the public API.

```kotlin
  fun listen(): Observable<AppUpdateState> {
    return appUpdateCallback()
        .flatMap(this::shouldNudgeForUpdate)
        .onErrorReturn(::AppUpdateStateError)
  }

  private fun appUpdateCallback(): Observable<AppUpdateInfo> {
    val appUpdateManager = AppUpdateManagerFactory.create(appContext)
    val appUpdateInfoTask = appUpdateManager.appUpdateInfo

    return Observable.create { emitter ->
      appUpdateInfoTask.addOnSuccessListener { appUpdateInfo ->
        emitter.onNext(appUpdateInfo)
      }

      appUpdateInfoTask.addOnFailureListener { exception ->
        emitter.onError(exception)
      }
    }
  }
```

What do you think?

Reply:
I felt the same about using `sealed class` here but I didn't think of `onErrorReturn` as an option. Thanks!
Will make this change.